include/irr/asset/IImageAssetHandlerBase.h

```diff
@@ -31,10 +37,207 @@ class IImageAssetHandlerBase : public virtual core::IReferenceCounted
+
+				auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1);
+				newRegions->front() = *referenceTopRegion;
+
+				const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(referenceImageParams.format);
+				auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texelOrBlockByteSize * newRegions->front().bufferRowLength * newImageParams.extent.height * newImageParams.extent.depth);
+
+				newImage = ICPUImage::create(std::move(newImageParams));
+				newImage->setBufferAndRegions(std::move(texelBuffer), newRegions);
+			}
+
+			using COPY_FILTER = CCopyImageFilter;
+			COPY_FILTER copyFilter;
+			COPY_FILTER::state_type state;
+
+			auto newTopRegion = newImage->getRegions().begin();
+
+			state.inImage = referenceImage.get();
+			state.outImage = newImage.get();
+			state.inOffset = { 0, 0, 0 };
+			state.outOffset = { 0, 0, 0 };
+			state.inBaseLayer = 0;
+			state.outBaseLayer = 0;
+			state.extent = newTopRegion->getExtent();
+			state.layerCount = 1;
+			state.inMipLevel = newTopRegion->imageSubresource.mipLevel;
+			state.outMipLevel = 0;
+
+			if(!copyFilter.execute(&state))
+				os::Printer::log("Something went wrong while copying top level region texel's data to the image!", ELL_WARNING);
+
+			return newImage;
+		}
+
+		/*
+			Create an image containing a single row from taking a raw data 
+			to a single row and convert it to any format. Since it's
+			data may not only limit to stuff being displayed on a screen,
+			there is an optiomal parameter for bufferRowLength pitch that
+			is helpful while dealing with specific data which needs it.
+		*/
+
+		template<E_FORMAT inputFormat, E_FORMAT outputFormat>
+		static inline core::smart_refctd_ptr<ICPUImage> createSingleRowImageFromRawData(void* rowData, uint32_t texelOrBlockLength, bool createWithBufferRowLengthPitch = false)
+		{
+			using CONVERSION_FILTER = CConvertFormatImageFilter<inputFormat, outputFormat>;
+			CONVERSION_FILTER convertFilter;
+			CONVERSION_FILTER::state_type state;
+
+			auto createImage = [&](E_FORMAT format, bool copyInputMemory = true)
+			{
+				const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(format);
+				const uint32_t pitchTexelOrBlockLength = createWithBufferRowLengthPitch ? calcPitchInBlocks(texelOrBlockLength, texelOrBlockByteSize) : texelOrBlockLength;
+
+				ICPUImage::SCreationParams imgInfo;
+				imgInfo.format = format;
+				imgInfo.type = ICPUImage::ET_1D;
+				imgInfo.extent = { texelOrBlockLength * asset::getBlockDimensions(format).X, 1, 1 };
+				imgInfo.mipLevels = 1u;
+				imgInfo.arrayLayers = 1u;
+				imgInfo.samples = ICPUImage::ESCF_1_BIT;
+				imgInfo.flags = static_cast<IImage::E_CREATE_FLAGS>(0u);
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+				auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texelOrBlockByteSize * pitchTexelOrBlockLength);
```

Comment:
you could have already passed this in, instead of `rowData`

Reply:
okay, so I will exchange void* to ICPUBuffer

Reply:
use smart pointers and move semantics

Reply:
there is no memcpy